src/Repository/RoadmapRepository.php

```diff
@@ -47,4 +48,20 @@ public function findOneBySomeField($value): ?Roadmap
         ;
     }
     */
+
+    /**
+     * @return Roadmap|null
+     */
+    public function getRoadmap(int $roadmapId): ?Roadmap
+    {
+        $roadmap = $this->createQueryBuilder('r')
+            ->join('r.steps', 's')
+            ->join('s.actions', 'a')
```

Comment:
Only add joins won't make it better than easyloading, you need to use "addSelect" to include the steps and actions fields in the query